spec/requests/graphql/mutations/create_asset_spec.rb

```diff
@@ -0,0 +1,100 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+require 'digest'
+
+RSpec.describe Mutations::CreateAsset, type: :request do
+  let(:authorized_object) { FactoryBot.create(:item, :with_primary_project) }
+  let(:authorized_project) { authorized_object.projects.first }
+  let(:blob_content) { "This is text to store in a blob" }
+  let(:blob_checksum) { Digest::MD5.hexdigest blob_content }
+  let(:blob_args) do
+    {
+      filename: 'blob.xyz',
+      byte_size: blob_content.bytesize,
+      checksum: blob_checksum,
+      content_type: 'text/plain'
+    }
+  end
+  let(:active_storage_blob) do
+    blob = ActiveStorage::Blob.create_before_direct_upload!(**blob_args)
+    blob.upload(StringIO.new(blob_content))
+    blob
+  end
+
+  include_examples 'requires user to have correct permissions for graphql request' do
+    let(:variables) do
+      {
+        input: {
+          parentId: authorized_object.uid, signedBlobId: 'not-relevant'
+        }
+      }
+    end
+    let(:request) { graphql query, variables }
+  end
+
+  context 'when logged in user is an administrator' do
+    before { sign_in_user as: :administrator }
+
+    context 'when adding asset from an upload' do
+      let(:variables) do
+        {
+          input: {
+            parentId: authorized_object.uid, signedBlobId: active_storage_blob.signed_id
+          }
+        }
+      end
+
+      before { graphql query, variables }
+
+      it 'returns a new asset' do
+        expect(response.body).to have_json_type(String).at_path('data/createAsset/asset/id')
+      end
+      it 'deletes the upload' do
+        expect(ActiveStorage::Blob.exists?(active_storage_blob.id)).to be false
+      end
+    end
+    context 'when asset save fails' do
+      let(:variables) do
+        {
+          input: {
+            parentId: authorized_object.uid, signedBlobId: active_storage_blob.signed_id
+          }
+        }
+      end
+      let(:predictable_file) { Tempfile.new(['blob', '.xyz']) }
+      let(:predictable_file_path) { predictable_file.path }
+      let(:storage_adapter) { Hyacinth::Config.resource_storage.primary_storage_adapter }
+      let(:predictable_location_uri) { storage_adapter.uri_prefix + predictable_file_path }
+      let(:expected_error_message) { "Testing upload deletes" }
+
+      before do
+        authorized_object
+        expect(Hyacinth::Config.resource_storage).to receive(:generate_new_location_uri).and_return(predictable_location_uri)
+        call_count = 0
+        allow(Hyacinth::Config.metadata_storage).to receive(:write) do |location, content|
+          call_count += 1
+          raise expected_error_message if call_count > 1
+          original_write.call(location, content)
```

Comment:
@barmintor Where is `original_write` defined?

Reply:
you got it.

Reply:
oh, wait - is it? But if it's not, why does this test run is the question.

Reply:
I don't think this can be in a request spec, not without doing very questionable things. I think we should remove this commit from the PR and open a ticket for unit testing the graphql resolvers.

Reply:
"not without doing very questionable things" or digging into the design of digital object concerns and storage adapters, which I think is out of scope of the ticket.

Reply:
I'm pushing a non-request spec that uses mocks and isolates the mutation.

Reply:
Yeah, I agree.  Ticketing and handling outside of PR sounds good.  Also, maybe we should consider moving the cleanup stuff into the model if possible, rather than the graphql mutation.